service/dap/types.go

```diff
@@ -97,20 +128,65 @@ func (m *SubstitutePath) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
-// AttachConfig is the collection of attach request attributes recognized by delve DAP implementation.
-type AttachConfig struct {
-	// Backend used by delve. Maps to `dlv`'s `--backend` flag.
-	// Allowed Values: `"default"`, `"native"`, `"lldb"`
-	Backend string `json:"backend,omitempty"`
+type AttachMode string
 
-	// Absolute path to the working directory of the program being debugged.
-	Cwd string `json:"cwd,omitempty"`
+// AttachMode is the type of an attach mode.
+const (
+	// "": same as LocalAttachMode
+	DefaultAttachMode AttachMode = ""
+	// "local": attaches to the local process with the given ProcessID. This is the default mode.
+	LocalAttachMode AttachMode = "string"
+)
 
-	// Only empty or `local` is acceptable.
-	Mode string `json:"mode,omitempty"`
+func (m *AttachMode) UnmarshalJSON(data []byte) error {
+	var mode string
+	if err := json.Unmarshal(data, &mode); err != nil {
+		if _, ok := err.(*json.UnmarshalTypeError); ok {
+			return fmt.Errorf(`cannot unmarshal '%s' into 'mode' of type string`, data)
+		}
+		return err
+	}
+	switch mode {
+	case "", "local":
+		*m = AttachMode(mode)
+	default:
+		return fmt.Errorf(`unsupported 'mode' value %q`, mode)
+	}
+	return nil
+}
 
-	// The numeric ID of the process to be debugged. Must not be 0.
+// AttachConfig is the collection of attach request attributes recognized by delve DAP implementation.
+type AttachConfig struct {
+	// Attach mode.
+	Mode AttachMode `json:"mode,omitempty"`
+
+	// The numeric ID of the process to be debugged. Required and must not be 0.
 	ProcessID int `json:"processId,omitempty"`
 
 	LaunchAttachCommonConfig
 }
+
+func unmarshalLaunchAttachArgs(input map[string]interface{}, config interface{}) error {
+	// TODO(hyangah): after go-dap PR 57 is merged, update the dependency and change to use
+	// the new API.
+	//   func unmarshalLaunchAttachArgs(input json.RawMessage, config interface{}) error { ... }
+	buf := &bytes.Buffer{}
+	if err := json.NewEncoder(buf).Encode(input); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(buf.Bytes(), config); err != nil {
+		if uerr, ok := err.(*json.UnmarshalTypeError); ok {
+			// Format json.UnmarshalTypeError error string in our own way. E.g.,
+			//   "json: cannot unmarshal number into Go struct field LaunchArgs.substitutePath of type dap.SubstitutePath"
+			//   => "cannot unmarshal number into 'substitutePath' of type {from:string, to:string}"
+			//   "json: cannot unmarshal number into Go struct field LaunchArgs.program of type string" (go1.16)
+			//   => "cannot unmarshal number into 'program' of type string"
+			if strings.HasPrefix(uerr.Value, "substitutePath") {
```

Comment:
Why has prefix instead of full mach?

Reply:
Ooops. I meant uerr.Field, not uerr.Value here.

Changed to the exact matching. Before implementing SubstitutePath custom unmarshaller, encoding/json may report 'substitutePath.from' if decoding error occurs while processing 'from'.